File: Raven.Client.Java/src/main/java/net/ravendb/abstractions/indexing/SpatialOptions.java
package net.ravendb.abstractions.indexing;

import net.ravendb.abstractions.basic.UseSharpEnum;

import org.apache.commons.lang.builder.HashCodeBuilder;


public class SpatialOptions {

  //about 4.78 meters at equator, should be good enough (see: http://unterbahn.com/2009/11/metric-dimensions-of-geohash-partitions-at-the-equator/)
  public final static int DEFAULT_GEOHASH_LEVEL = 9;

  //about 4.78 meters at equator, should be good enough
  public final static int DEFAULT_QUAD_TREE_LEVEL = 23;

  public SpatialOptions() {
    type = SpatialFieldType.GEOGRAPHY;
    strategy = SpatialSearchStrategy.GEOHASH_PREFIX_TREE;
    maxTreeLevel = DEFAULT_GEOHASH_LEVEL;
    minX = -180;
    maxX = 180;
    minY = -90;
    maxY = 90;
    units = SpatialUnits.KILOMETERS;
  }

  private SpatialFieldType type = SpatialFieldType.GEOGRAPHY;
  private SpatialSearchStrategy strategy = SpatialSearchStrategy.GEOHASH_PREFIX_TREE;
  private int maxTreeLevel;
  private double minX;
  private double maxX;
  private double minY;
  private double maxY;

  // Circle radius units, only used for geography  indexes
  private SpatialUnits units;



  public SpatialFieldType getType() {
    return type;
  }

  public void setType(SpatialFieldType type) {
    this.type = type;
  }

  public SpatialSearchStrategy getStrategy() {
    return strategy;
  }

  public void setStrategy(SpatialSearchStrategy strategy) {
    this.strategy = strategy;
  }

  public int getMaxTreeLevel() {
    return maxTreeLevel;
  }

  public void setMaxTreeLevel(int maxTreeLevel) {
    this.maxTreeLevel = maxTreeLevel;
  }

  public double getMinX() {
    return minX;
  }

  public void setMinX(double minX) {
    this.minX = minX;
  }

  public double getMaxX() {
    return maxX;
  }

  public void setMaxX(double maxX) {
    this.maxX = maxX;
  }

  public double getMinY() {
    return minY;
  }

  public void setMinY(double minY) {
    this.minY = minY;
  }

  public double getMaxY() {
    return maxY;
  }

  public void setMaxY(double maxY) {
    this.maxY = maxY;
  }

  public SpatialUnits getUnits() {
    return units;
  }

  public void setUnits(SpatialUnits units) {
    this.units = units;
  }


  @Override
  public boolean equals(Object obj) {
    if (this == obj)
      return true;
    if (obj == null)
      return false;
    if (getClass() != obj.getClass())
      return false;
    SpatialOptions other = (SpatialOptions) obj;

    boolean result = type == other.getType() && strategy == other.strategy;
    if (type == SpatialFieldType.GEOGRAPHY) {
      result = result && units == other.units;
    }
    if (strategy != SpatialSearchStrategy.BOUNDING_BOX) {
      result = result && maxTreeLevel == other.maxTreeLevel;
      if (type == SpatialFieldType.CARTESIAN) {
        result = result && minX == other.minX
            && maxX == other.maxX
            && minY == other.minY
            && maxY == other.maxY;
      }
    }

    return result;
  }

  @Override
  public int hashCode() {
    HashCodeBuilder builder = new HashCodeBuilder();
    builder.append(type);
    builder.append(strategy);
    if (type == SpatialFieldType.GEOGRAPHY) {
      builder.append(units.hashCode());
    }
    if (strategy != SpatialSearchStrategy.BOUNDING_BOX) {
      builder.append(maxTreeLevel);
      if (type == SpatialFieldType.CARTESIAN) {
        builder.append(minX).append(maxX).append(minY).append(maxY);
      }
    }


    return builder.hashCode();
  }

  @UseSharpEnum
  public enum SpatialFieldType {
    GEOGRAPHY, CARTESIAN
  }

  @UseSharpEnum
  public enum SpatialSearchStrategy {
    GEOHASH_PREFIX_TREE, QUAD_PREFIX_TREE, BOUNDING_BOX
  }

  @UseSharpEnum
  public enum SpatialRelation {
    WITHIN, CONTAINS, DISJOINT, INTERSECTS,

    /**
     * Does not filter the query, merely sort by the distance
     */
    NEARBY
  }

  @UseSharpEnum
  public enum SpatialUnits {
    KILOMETERS, MILES
  }

}

File: src/Raven.Client/Documents/Indexes/Spatial/SpatialOptions.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Raven.Client.Documents.Indexes.Spatial
{
    public class AutoSpatialOptions : SpatialOptions
    {
        public AutoSpatialMethodType MethodType { get; set; }

        public List<string> MethodArguments { get; set; }

        public AutoSpatialOptions()
        {
            MethodArguments = new List<string>();
        }

        public AutoSpatialOptions(AutoSpatialMethodType methodType, List<string> methodArguments)
        {
            MethodType = methodType;
            MethodArguments = methodArguments ?? throw new ArgumentNullException(nameof(methodArguments));
        }

        public AutoSpatialOptions(AutoSpatialOptions options)
            : base(options)
        {
            MethodType = options.MethodType;
            MethodArguments = new List<string>(options.MethodArguments);
        }

        protected bool Equals(AutoSpatialOptions other)
        {
            return base.Equals(other) && MethodType == other.MethodType && MethodArguments.SequenceEqual(other.MethodArguments);
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj))
                return false;
            if (ReferenceEquals(this, obj))
                return true;
            if (obj.GetType() != GetType())
                return false;
            return Equals((AutoSpatialOptions)obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                int hashCode = base.GetHashCode();
                hashCode = (hashCode * 397) ^ (int)MethodType;
                hashCode = (hashCode * 397) ^ (MethodArguments != null ? MethodArguments.GetHashCode() : 0);
                return hashCode;
            }
        }

        public enum AutoSpatialMethodType
        {
            Point,
            Wkt
        }
    }

    public class SpatialOptions
    {
        // about 4.78 meters at equator, should be good enough (see: http://unterbahn.com/2009/11/metric-dimensions-of-geohash-partitions-at-the-equator/)
        public const int DefaultGeohashLevel = 9;

        // about 4.78 meters at equator, should be good enough
        public const int DefaultQuadTreeLevel = 23;

        internal static readonly SpatialOptions Default = new SpatialOptions();

        public SpatialOptions()
        {
            Type = SpatialFieldType.Geography;
            Strategy = SpatialSearchStrategy.GeohashPrefixTree;
            MaxTreeLevel = DefaultGeohashLevel;
            MinX = -180;
            MaxX = 180;
            MinY = -90;
            MaxY = 90;
            Units = SpatialUnits.Kilometers;
        }

        internal SpatialOptions(SpatialOptions options)
        {
            Type = options.Type;
            Strategy = options.Strategy;
            MaxTreeLevel = options.MaxTreeLevel;
            MinX = options.MinX;
            MaxX = options.MaxX;
            MinY = options.MinY;
            MaxY = options.MaxY;
            Units = options.Units;
        }

        public SpatialFieldType Type { get; set; }
        public SpatialSearchStrategy Strategy { get; set; }
        public int MaxTreeLevel { get; set; }
        public double MinX { get; set; }
        public double MaxX { get; set; }
        public double MinY { get; set; }
        public double MaxY { get; set; }

        /// <summary>
        /// Circle radius units, only used for geography indexes
        /// </summary>
        public SpatialUnits Units { get; set; }

        protected bool Equals(SpatialOptions other)
        {
            var result = Type == other.Type && Strategy == other.Strategy;

            if (Type == SpatialFieldType.Geography)
            {
                result = result && Units == other.Units;
            }

            if (Strategy != SpatialSearchStrategy.BoundingBox)
            {
                result = result && MaxTreeLevel == other.MaxTreeLevel;

                if (Type == SpatialFieldType.Cartesian)
                {
                    result = result
                        && MinX.Equals(other.MinX)
                        && MaxX.Equals(other.MaxX)
                        && MinY.Equals(other.MinY)
                        && MaxY.Equals(other.MaxY);
                }
            }
            return result;
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj))
                return false;
            if (ReferenceEquals(this, obj))
                return true;
            if (obj.GetType() != GetType())
                return false;
            return Equals((SpatialOptions)obj);
        }

        public override int GetHashCode()
        {
            unchecked
            {
                int hashCode = (int)Type;
                hashCode = (hashCode * 397) ^ (int)Strategy;

                if (Type == SpatialFieldType.Geography)
                {
                    hashCode = (hashCode * 397) ^ Units.GetHashCode();
                }

                if (Strategy != SpatialSearchStrategy.BoundingBox)
                {
                    hashCode = (hashCode * 397) ^ MaxTreeLevel;

                    if (Type == SpatialFieldType.Cartesian)
                    {
                        hashCode = (hashCode * 397) ^ MinX.GetHashCode();
                        hashCode = (hashCode * 397) ^ MaxX.GetHashCode();
                        hashCode = (hashCode * 397) ^ MinY.GetHashCode();
                        hashCode = (hashCode * 397) ^ MaxY.GetHashCode();
                    }
                }

                return hashCode;
            }
        }
    }

    public enum SpatialFieldType
    {
        Geography,
        Cartesian
    }

    public enum SpatialSearchStrategy
    {
        GeohashPrefixTree,
        QuadPrefixTree,
        BoundingBox
    }

    public enum SpatialRelation
    {
        Within,
        Contains,
        Disjoint,
        Intersects
    }

    public enum SpatialUnits
    {
        Kilometers,
        Miles
    }
}
